src/ngraph/distributed.cpp

```diff
@@ -14,90 +14,35 @@
 // limitations under the License.
 //*****************************************************************************
 
-#ifdef NGRAPH_DISTRIBUTED_ENABLE
-
-#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
-#include <mlsl.hpp>
-#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
-#include <mpi.h>
-#endif
-
 #include "ngraph/distributed.hpp"
+#include "ngraph/distributed/mlsl.hpp"
+#include "ngraph/distributed/null.hpp"
+#include "ngraph/distributed/open_mpi.hpp"
 #include "ngraph/log.hpp"
 
 using namespace ngraph;
 
-ngraph::Distributed::Distributed()
-{
-#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
-    if (!MLSL::Environment::GetEnv().IsInitialized())
-    {
-        MLSL::Environment::GetEnv().Init(nullptr, nullptr);
-        m_init_comm = true;
-    }
-#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
-    int flag = 0;
-    MPI_Initialized(&flag);
-    if (!flag)
-    {
-        MPI_Init(NULL, NULL);
-        m_init_comm = true;
-    }
-#else
-    throw ngraph_error("Distributed Library not supported/mentioned");
-#endif
-}
+static std::unique_ptr<DistributedInterface> s_distributed_interface;
 
-ngraph::Distributed::~Distributed()
+void ngraph::set_distributed_interface(std::unique_ptr<DistributedInterface> distributed_interface)
 {
-    if (m_init_comm == true)
-    {
-        finalize();
-    }
+    s_distributed_interface = std::move(distributed_interface);
 }
 
-void ngraph::Distributed::finalize()
+DistributedInterface* ngraph::get_distributed_interface()
 {
-#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
-    if (MLSL::Environment::GetEnv().IsInitialized())
-    {
-        MLSL::Environment::GetEnv().Finalize();
-    }
-#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
-    int flag = 0;
-    MPI_Initialized(&flag);
-    if (flag)
+    if (0 == s_distributed_interface)
     {
-        MPI_Finalize();
-    }
-#else
-    throw ngraph_error("Distributed Library not supported/mentioned");
-#endif
-}
-
-int ngraph::Distributed::get_size() const
-{
-#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
-    return static_cast<int>(MLSL::Environment::GetEnv().GetProcessCount());
-#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
-    int size;
-    MPI_Comm_size(MPI_COMM_WORLD, &size);
-    return size;
+#ifdef NGRAPH_DISTRIBUTED_OMPI_ENABLE
+        set_distributed_interface(std::unique_ptr<DistributedInterface>(
```

Comment:
could you please put some log here and below to show when this is set?

Reply:
Added a settable name to the interfaces and logged setting of the interface.